### PR TITLE
Fix bug for RestoreSnapshot()

### PR DESCRIPTION
**Pull request by zhrt123**
This PR fix bug: snapshot is uninitialized after allocation:
- haveDistribSnapshot
- speculativeToken
- ph_node

In my investigation, `haveDistribSnapshot` is used without initialization.
For example, in parallel worker (we want to apply parallel worker for parallel index creation in the next work):
```
	asnapspace = shm_toc_lookup(toc, PARALLEL_KEY_ACTIVE_SNAPSHOT, false);
	tsnapspace = shm_toc_lookup(toc, PARALLEL_KEY_TRANSACTION_SNAPSHOT, true);
	asnapshot = RestoreSnapshot(asnapspace);
	tsnapshot = tsnapspace ? RestoreSnapshot(tsnapspace) : asnapshot;
	RestoreTransactionSnapshot(tsnapshot,
							   fps->parallel_leader_pgproc);
```
The snapshot will be used in RestoreTransactionSnapshot(), and the field `haveDistribSnapshot` of Snapshot will be used without initialization.

### CODE REVIEW

File: src/backend/utils/time/snapmgr.c [hunk 1]
```diff
@@ -2384,6 +2384,8 @@ RestoreSnapshot(char *start_address)
 
 	/* Copy all required fields */
 	snapshot = (Snapshot) MemoryContextAlloc(TopTransactionContext, size);
+	memset(snapshot, 0, sizeof(SnapshotData));
```
Comment (adam8157) [inline src/backend/utils/time/snapmgr.c:2387]: `MemoryContextAllocZero()`

**Comment by haolinw** [comment]
Is that GPDB specific change for Upstream base code ? If so, please add comments such as "GPDB specific code ..." to clarify the purpose.

**Review by adam8157 — CHANGES_REQUESTED** [review]
These variables should be stored into and restored from `serialized_snapshot`, setting them to 0/false lost the information.

**Comment by zhrt123** [comment]
> Is that GPDB specific change for Upstream base code ? If so, please add comments such as "GPDB specific code ..." to clarify the purpose.

I will do it.

File: src/backend/utils/time/snapmgr.c (hunk 1 above)
Comment (zhrt123) [reply src/backend/utils/time/snapmgr.c:2387]: This may create conflict with upstream, do we need to make this change?

**Comment by zhrt123** [comment]
> These variables should be stored into and restored from `serialized_snapshot`, setting them to 0/false lost the information.

hmm, we just set them to 0 when allocating the memory. Shall we store them when serializing snapshots?

**Comment by kainwen** [comment]
Reading the commit message, it seems this PR is only for parallel workers. Given the fact that we still don't decide to fully support parallel workers yet. I don't think we should ship this as a separate PR.

Or, would you please create a case that current GPDB can hit the bug?

**Comment by zhrt123** [comment]
> Reading the commit message, it seems this PR is only for parallel workers. Given the fact that we still don't decide to fully support parallel workers yet. I don't think we should ship this as a separate PR.
> 
> Or, would you please create a case that current GPDB can hit the bug?

Not yet. But it still is a bug in a general function. Engineers who use this function may be affected and they need to find the bug and re-fix it.

**Comment by kainwen** [comment]
> > Reading the commit message, it seems this PR is only for parallel workers. Given the fact that we still don't decide to fully support parallel workers yet. I don't think we should ship this as a separate PR.
> > Or, would you please create a case that current GPDB can hit the bug?
> 
> Not yet. But it still is a bug in a general function. Engineers who use this function may be affected and they need to find the bug and re-fix it.

Is it possible to create a case that trigger this BUG?

**Comment by zhrt123** [comment]
> > > Reading the commit message, it seems this PR is only for parallel workers. Given the fact that we still don't decide to fully support parallel workers yet. I don't think we should ship this as a separate PR.
> > > Or, would you please create a case that current GPDB can hit the bug?
> > 
> > 
> > Not yet. But it still is a bug in a general function. Engineers who use this function may be affected and they need to find the bug and re-fix it.
> 
> Is it possible to create a case that trigger this BUG?

hmm, currently no scene can trigger this bug in gpdb... We need to create a bgworker and pass the snapshot to this bgworker. Shall we add some C code to apply this test?

**Comment by kainwen** [comment]
> > > > Reading the commit message, it seems this PR is only for parallel workers. Given the fact that we still don't decide to fully support parallel workers yet. I don't think we should ship this as a separate PR.
> > > > Or, would you please create a case that current GPDB can hit the bug?
> > > 
> > > 
> > > Not yet. But it still is a bug in a general function. Engineers who use this function may be affected and they need to find the bug and re-fix it.
> > 
> > 
> > Is it possible to create a case that trigger this BUG?
> 
> Currently, no scene can trigger this bug in gpdb... We need to create a bgworker and pass the snapshot to this bgworker. Shall we add some C code to apply this test? @kainwen

GDD and auto analyze (auto vacuum) will create bgworker, would you please  try to help verify those? @zhrt123

File: src/backend/utils/time/snapmgr.c (hunk 1 above)
Comment (adam8157) [reply src/backend/utils/time/snapmgr.c]: Upstream restores the snapshot with `speculativeToken` and `ph_node` uninitialized, worth submitting a patch.

**Comment by adam8157** [comment]
The `pg_export_snapshot()/ExportSnapshot()` handles `haveDistribSnapshot` well, I agree `SerializeSnapshot()` and `RestoreSnapshot()` generally should do that too.

It's just better to have a thing that proves it works, suggest having a case or converting this PR to a separate commit of your WIP parallel-worker-related PR.

**Pull request closed without merging**